spec/services/moves/sweeper_spec.rb

```diff
@@ -54,117 +56,105 @@
-            from_location: wood_green_court,
-            to_location: Location.find_by(nomis_agency_id: attributes[:to_location_nomis_agency_id])
-          )
-        end
-
-        it 'cancels one move' do
-          expect { sweeper.call }.to change { Move.where(status: :cancelled).count }.by(1)
-        end
-
-        it 'cancels the move that did match' do
-          sweeper.call
-          expect(outdated_move.reload.status).to eq 'cancelled'
-        end
-
-        it 'leaves the move that did not match by location' do
-          sweeper.call
-          expect(court_move.reload.status).to eq 'requested'
-        end
+      before do
+        Move.create!(
+          date: '2019-09-15',
+          from_location: brixton_prison,
+          to_location: brixton_prison,
+          person: prisoner_two,
+          status: 'requested',
+          time_due: '2019-08-19 08:00:00',
+          nomis_event_ids: [487_463_210]
+        )
       end
 
-      context 'with multiple locations' do
-        let(:locations) { [brixton_prison, wood_green_court] }
-        let!(:court_move) do
-          Move.create!(
-            date: attributes[:date],
-            time_due: attributes[:time_due],
-            nomis_event_ids: [487_463_208],
-            person: Person.find_by(nomis_prison_number: attributes[:person_nomis_prison_number]),
-            from_location: wood_green_court,
-            to_location: Location.find_by(nomis_agency_id: attributes[:to_location_nomis_agency_id])
-          )
-        end
-
-        it 'cancels two moves' do
-          expect { sweeper.call }.to change { Move.where(status: :cancelled).count }.by(2)
-        end
-
-        it 'cancels the move that matched the first location' do
-          sweeper.call
-          expect(outdated_move.reload.status).to eq 'cancelled'
-        end
-
-        it 'cancels the move that matched the second location' do
-          sweeper.call
-          expect(court_move.reload.status).to eq 'cancelled'
-        end
+      it 'empties the nomis_event_ids array' do
+        sweeper.call
+        expect(outdated_move.reload.nomis_event_ids).to be_empty
       end
 
-      context 'with a move with multiple nomis_event_ids' do
-        let!(:multiple_nomis_ids_move) do
-          Move.create!(
-            date: attributes[:date],
-            time_due: attributes[:time_due],
-            nomis_event_ids: [attributes[:nomis_event_id], 123_456_789],
-            person: prisoner_one,
-            from_location: brixton_prison,
-            to_location: wood_green_court
-          )
-        end
-
-        it 'removes the outdated nomis_event_id' do
-          sweeper.call
-          expect(multiple_nomis_ids_move.reload.nomis_event_ids).to eq([attributes[:nomis_event_id]])
-        end
+      it 'cancels the outdated move' do
+        sweeper.call
+        expect(outdated_move.reload.status).to eq('cancelled')
       end
     end
   end
```

Comment:
Do we need an additional test case where there are no NOMIS ids? I may have missed this one...

Reply:
I thought it was implicit that that case was covered, because the one you're commenting works (id gets removed from the array, then state changes to cancelled because the array is empty). Makes sense?

Reply:
Sorry, I meant for moves that never belonged to NOMIS (i.e. the data we lost before)

Reply:
There's no clear way to identify that, they all have empty nomis_event_ids. We have same move from nomis and outside of it now?

Reply:
There is a really clear way to identify locations that should be synced with NOMIS. If the NOMIS client is hitting the **movements** api requesting moves from.. lets say HMP Pentonville (PVI) then this site and all the moves originating from it, should be synced. The movements api is key as this tells us users are not entering moves in our service for that site.(they also have the pecs_prison role & caseload - our workflow doesn't allow us create moves in our service) What sites a user seeing is due to HMPPS Authentication (If it is an Auth user (outside of NOMIS) its a police station, If its non Auth (inside NOMIS - we have to do a secondary lookup to get caseload) then its a Prison). Not all requests from the frontend invoke the NOMIS lookup - all that logic is in the codebase.

Reply:
:+1: @ngw this might be what you're after: https://github.com/ministryofjustice/hmpps-book-secure-move-api/blob/master/app/models/location.rb#L21-L23

Reply:
Ok, but then the logic we currently have is 100% correct and covering all the cases, right? The sweeper does not act on "all the moves", but limited to moves that are passed as arguments. Once the cancellation of the moves is scoped on the locations provided (which are provided directly by nomis) we should be perfectly safe.

Reply:
Hi @ngw, I think we should just merge it at this point, but your comment above makes me feel quite uneasy. We should be coding defensively, particularly on this sort of thing where important data is being modified. The assumption that it's "100% correct" relies on the logic in another class not changing, which means we have a dependency on the thing that invokes this... it's easy to foresee this being called somewhere else in future, and without that defensive check we _could_ alter data in a way we might not expect. 

It's good to be kind to our future selves and protect ourselves from that sort of error.

As an example, you could write a test case here that invokes the sweeper for a location which is not a prison and has no `nomis_event_ids`. Even if we don't add that test before merging this (and I believe @jason-sims is keen to manually test this in staging today), I'd like to see this test in a subsequent PR.